### DISCUSSION TITLE
How to use sub threads to update the text of a Label?

### DISCUSSION THREAD

**Discussion by chengw-cn** [beeware/toga#1883] [Q&A]
How to use sub threads to update the text of a Label / value of MultilineTextInput?
I want to use LABEL to display constantly updated text.
look at my code,
Press Record Button OK, press Connect Button failed!
help!  Thank you very much.

```
import time
import threading
import socket
import toga
from toga.style import Pack
from toga.style.pack import COLUMN, ROW,LEFT,RIGHT,CENTER
import random


class HelloWorld(toga.App):

    def __init__(
        self,
        formal_name=None,
        app_id=None,
        app_name=None,
        id=None,
        icon=None,
        author=None,
        version=None,
        home_page=None,
        description=None,
        startup=None,
        windows=None,
        on_exit=None,
        factory=None  # DEPRECATED !
        ):
        super().__init__(
            formal_name,
            app_id,
            app_name,
            id,
            icon,
            author,
            version,
            home_page,
            description,
            startup,
            windows,
            on_exit,
            factory
            )
        self.c_input = None
        self.f_input = None
        self.show_label = None
        self.sta_label = None
        self.textbox = None

    def fun_test(self):
        print("test ...")
        while True:
            print("test ...11")
            # self.textbox.value = str(random.randint(1000,9999))
            self.sta_label.text = str(random.randint(1000,9999)) 
            self.show_label.text = str(random.randint(1000,9999))
            time.sleep(0.2)

    def fun_conn(self,widget):
        print('fun_conn ...')
        th=threading.Thread(target=self.fun_test,args=())
        th.setDaemon(True)
        th.start()

    def fun_rec(self,widget):
        print('fun_rec ...')
        self.textbox.value = str(random.randint(1000,9999))
        self.sta_label.text = str(random.randint(1000,9999)) 
        self.show_label.text = str(random.randint(1000,9999))

    def startup(self):
        """
        Construct and show the Toga application.
        Usually, you would add your application to a main content box.
        We then create a main window (with a name matching the app), and
        show the main window.
        """
        main_box = toga.Box()

        print('startup ...')
        c_box = toga.Box()
        f_box = toga.Box()
        d_box = toga.Box()

        
        self.c_input = toga.TextInput(value='11.22.205.155:60660')#(readonly=True)
        self.f_input = toga.TextInput(value='60661')

        c_label = toga.Label("ServerAdd", style=Pack(text_align=LEFT))
        f_label = toga.Label("RTK Port", style=Pack(text_align=LEFT))
        # join_label = toga.Label("", style=Pack(text_align=RIGHT))
        
        self.show_label = toga.Label(text="No data ...", style=Pack(height=100,background_color='#E0E0E0'))
        self.sta_label = toga.Label(text='SEISYS',style=Pack(width=150,color='#0080FF',text_align=CENTER))
        self.textbox = toga.MultilineTextInput(value='',readonly=True,style=Pack(height=220))
        
        bt_con = toga.Button("Connect", on_press=self.fun_conn)
        bt_rec = toga.Button("Record", on_press=self.fun_rec)
        
        c_box.add(c_label)
        c_box.add(self.c_input)
        # c_box.add(join_label)

        f_box.add(f_label)
        f_box.add(self.f_input)
        f_box.add(bt_con)
        d_box.add(self.sta_label)
        d_box.add(bt_rec)
        
        main_box.add(c_box)
        main_box.add(f_box)
        main_box.add(self.show_label)
        main_box.add(d_box)
        main_box.add(self.textbox)

        main_box.style.update(direction=COLUMN, padding=10)
        f_box.style.update(direction=ROW, padding=5)
        c_box.style.update(direction=ROW, padding=5)
        
        bt_rec.style.update(flex=1)
        self.c_input.style.update(flex=1)
        self.f_input.style.update(flex=0)
        c_label.style.update(width=120, padding_left=10)
        f_label.style.update(width=120, padding_left=10)
        # join_label.style.update(width=150, padding_right=10)
        
        d_box.style.update(padding=15)
        bt_con.style.update(flex=1,padding_left=30)
        
        self.main_window = toga.MainWindow(title=self.formal_name)
        self.main_window.content = main_box
        self.main_window.show()

def main():
    return HelloWorld('test','123')


if __name__ == "__main__":
    main().main_loop()
```

**Comment by freakboy3742** [comment]
The problem you're hitting is a general limitation of GUI frameworks. GUI frameworks require that you make any GUI modifications on the GUI thread - you can't modify a GUI from the non-GUI thread. In this case, you've set up a second thread so that you don't block the main GUI thread with the sleep calls - which is good - but you then can't make the label updates from the second thread. 

In most GUI frameworks, the fix for this is to post a message from the non-GUI worker thread to the GUI thread indicating the GUI update that you want to make. However, Toga has a much easier solution - a background task.

Redefine `def fun_test(self)` as `def fun_test(self, app, **kwargs)`, and change the `time.sleep(0.2)` to `yield 0.2`. Then, replace the implementation of `def fun_conn(self,widget)` with:
```
def fun_conn(self, widget):
     self.add_background_task(self.fun_test)
```

This will queue add `fun_test` as a background task; the value passed to the yield is the length of time that the worker thread will sleep. The background task will run on the main GUI thread, so it can update the GUI directly.

You can also use an async worker, by redefining `fun_test` as `async def fun_test(self, app, **kwargs)`, and using `asyncio.sleep(0.2)`. The implementation of `fun_conn` is the same - `add_background_task` can take either a generator or an async coroutine. If all you're doing is sleeping, the two approaches are the same; but the async coroutine approach also allows for any other async code you may need to do in the background (e.g., calling a network API).

**Reply by chengw-cn** [reply]
Thank you @freakboy3742 ! The problem has been resolved. I have completed my first Android APK using Python.
![apk](https://user-images.githubusercontent.com/131322435/233818823-34877f97-194a-484e-be10-2a1208c24b64.png)

But there's still a small issue, Setting MultilineTextInput style background_color is invalid in Android(Windows OK).
`self.textbox = toga.MultilineTextInput(value='',style=Pack(height=220,background_color='#E0E0E0')) #failed
self.textbox.style.update(background_color='#E0E0E0')#failed`
Can async be used on Android? I will give it a try.

`import time
import threading
import socket
import toga
from toga.style import Pack
from toga.style.pack import COLUMN, ROW,LEFT,RIGHT,CENTER
import random


class RTK_points_recorder(toga.App):

    def __init__(
        self,
        formal_name=None,
        app_id=None,
        app_name=None,
        id=None,
        icon=None,
        author=None,
        version=None,
        home_page=None,
        description=None,
        startup=None,
        windows=None,
        on_exit=None,
        factory=None  # DEPRECATED !
        ):
        super().__init__(
            formal_name,
            app_id,
            app_name,
            id,
            icon,
            author,
            version,
            home_page,
            description,
            startup,
            windows,
            on_exit,
            factory
            )
        self.c_input = None
        self.f_input = None
        self.show_label = None
        self.sta_label = None
        self.textbox = None
        self.rec_flag = 0
        self.conn_flag = 0
        self.rec_gps = ''
        self.x_input = None
    
    
    def fun_rec(self,widget):
        print('fun_rec ...')
        tt = time.time()
        if tt - self.rec_flag > 1:
            self.rec_flag = tt
            print('fun_rec ...ok')


    def fun_conn(self,widget):
        if not self.conn_flag:
            self.conn_flag = 1
            self.add_background_task(self.con_def)
        else:
            self.conn_flag = 0


    def con_def(self, app, **kwargs):
        print('con_def ...')
        s = socket.socket()         # 创建 socket 对象
        s.settimeout(0.1)
        host = self.c_input.value
        port = int(host.split(':')[1])    # 设置端口号
        host = host.split(':')[0]
        rtk_port = self.f_input.value
        # print('start ...',host,port,gns_port)
        try:
            s.connect((host, port))
            host = '$bd' + rtk_port
            s.send(host.encode('utf-8'))
            res = s.recv(120).decode()
            m = 0
            self.sta_label.text = 'ONLINE'
            while True:
                if self.conn_flag == 0:
                    self.sta_label.text = 'OFFLINE'
                    break
                m += 1
                if m >9:
                    m = 0
                res = ''
                try:
                    res = s.recv(1024).decode()
                    if '$GPGGA' in res: 
                        res = '$GPGGA' + res.split('$GPGGA')[1]
                        if not res.count(',') > 6:
                            continue
                        tem = res.split(',')
                        if not(tem[3] in 'NS' or tem[5] in 'EW'): 
                            continue
                        if not(tem[4].replace('.','').isdigit() or tem[2].replace('.','').isdigit()): 
                            continue
                        lo = str(float(tem[4][:tem[4].index('.')-2]) + float(tem[4][tem[4].index('.')-2:]) / 60)
                        if len(lo.split('.')[1]) > 8 : lo = lo.split('.')[0] + '.' + lo.split('.')[1][:8]
                        la = str(float(tem[2][:tem[2].index('.')-2]) + float(tem[2][tem[2].index('.')-2:]) / 60)
                        if len(la.split('.')[1]) > 8 : la = la.split('.')[0] + '.' + la.split('.')[1][:8]
                        # sta = res.split(',')[6]
                        # if sta == '0': sta = '(0)初始化'
                        # if sta == '1': sta = '(1)单点定位'
                        # if sta == '2': sta = '(2)码差分'
                        # if sta == '3': sta = '(3)无效PPS'
                        # if sta == '4': sta = '(4)固定解'
                        # if sta == '5': sta = '(5)浮点解'
                        # if sta == '6': sta = '(6)正在估算'
                        # if sta == '7': sta = '(7)人工输入固定值'
                        # if sta == '8': sta = '(8)模拟模式'
                        # if sta == '9': sta = '(9)WAAS差分'
                        res = '$' + lo + ',' + la + ',' + res.split(',')[6]

                    print('res:',res)
                    if res[-1:] == '4':
                        self.sta_label.text = res[-1:] + '[固定]'
                    elif res[-1:] == '1':
                        self.sta_label.text  = res[-1:] + '[单点]'
                    elif res[-1:] == '5':
                        self.sta_label.text  = res[-1:] + '[浮点]'
                    elif res[-1:] == '0':
                        self.sta_label.text = res[-1:] + '[初始化]'
                    else:
                        self.sta_label.text = res
                    res = res.replace('$','')
                    if ',' in res:
                        res = res.split(',')
                        self.show_label.text = res[0] + ',' +res[1] + ' [' +res[2] + ']  > ' + str(m)
                        if not self.rec_flag == 0:
                            if time.time() - self.rec_flag < 1.5:
                                # print('textbox.value = rec_gps ...')
                                self.rec_gps += '\n' + res[0] + ',' +res[1] + ',' +res[2] +','+ self.x_input.value
                                # print('aa',self.rec_gps,'bb')
                                self.textbox.value = self.rec_gps
                                self.rec_flag = 0
                                self.sta_label.text = 'Recorded.'
                            else:
                                self.sta_label.text = 'RecTimeOut'
                except (Exception,ValueError) as err:
                    er = str(err)
                    if not 'timed out' in er:
                        self.show_label.text = er
                        self.sta_label.text = 'Error.'
                    print(str(err))
                # time.sleep(0.5)
                yield 0.01

        except (Exception,ValueError) as err:
            print('连接错误：' + str(err))
            self.show_label.text = str(err)
            self.sta_label.text = 'Conn Failed'
        finally:
            s.close()


    def startup(self):
        main_box = toga.Box()

        print('startup ...')
        c_box = toga.Box()
        f_box = toga.Box()
        d_box = toga.Box()
        
        self.c_input = toga.TextInput(value='58.34.???.???:60660')#(readonly=True)
        self.f_input = toga.TextInput(value='60661')
        self.x_input = toga.TextInput(value='')

        c_label = toga.Label("ServerAdd:", style=Pack(text_align=LEFT))
        f_label = toga.Label("RTK Port:", style=Pack(text_align=LEFT))
        df_data = '   两次记点时间间隔大于1秒\n   按Record后1.5秒内如果无接收数据记点失败\n   下输入框为记点附加信息'
        self.show_label = toga.Label(text=df_data, style=Pack(height=80,background_color='#E0E0E0'))
        self.sta_label = toga.Label(text='SEISYS',style=Pack(width=120,color='#0080FF',text_align=LEFT))
        self.textbox = toga.MultilineTextInput(value='',style=Pack(height=220,background_color='#E0E0E0'))
        self.textbox.scroll_to_bottom()
        
        bt_con = toga.Button("Dis/connect", on_press=self.fun_conn)
        button = toga.Button("Record", on_press=self.fun_rec)
        
        c_box.add(c_label)
        c_box.add(self.c_input)

        f_box.add(f_label)
        f_box.add(self.f_input)
        f_box.add(bt_con)
        d_box.add(self.sta_label)
        d_box.add(self.x_input)
        d_box.add(button)
        
        main_box.add(c_box)
        main_box.add(f_box)
        main_box.add(self.show_label)
        main_box.add(d_box)
        main_box.add(self.textbox)

        main_box.style.update(direction=COLUMN, padding=5)
        f_box.style.update(direction=ROW, padding=5)
        c_box.style.update(direction=ROW, padding=5)
        self.x_input.style.update(padding=5,width=80)
        button.style.update(padding=5)
        
        button.style.update(flex=1)
        self.c_input.style.update(flex=1)
        self.f_input.style.update(flex=0,width=80)
        c_label.style.update(width=90, padding_left=10)
        f_label.style.update(width=90, padding_left=10)
        bt_con.style.update(flex=1,padding_left=5)
        self.textbox.style.update(flex=1)
        
        d_box.style.update(padding=10)
        self.main_window = toga.MainWindow(title=self.formal_name)
        self.main_window.content = main_box
        self.main_window.show()


def main():
    return RTK_points_recorder('RTK_points_recorder','1')

 if __name__ == "__main__":
     main().main_loop()
`

**Reply by freakboy3742** [reply]
There's no restriction on using async on Android.

The problem you're hitting is a bug - or, rather, a limitation of the MultilineTextInput on Android. Color change APIs aren't fully implemented on all platforms, on all widgets. We're in the process of auditing all our widgets with the hope of correcting this.

**Reply by chengw-cn** [reply]
Thank you @freakboy3742 !   I also hope that BeeWare will become better and better.
